omniscidb/QueryEngine/JoinHashTable/Runtime/HashJoinRuntimeCpu.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+
+#include "../../../Shared/sqltypes.h"
+
+namespace cpu_utils {
+/**
+ * Joins between two dictionary encoded string columns without a shared string dictionary
+ * are computed by translating the inner dictionary to the outer dictionary while filling
+ * the  hash table. The translation works as follows:
+ *
+ * Given two tables t1 and t2, with t1 the outer table and t2 the inner table, and two
+ * columns t1.x and t2.x, both dictionary encoded strings without a shared dictionary, we
+ * read each value in t2.x and do a lookup in the dictionary for t1.x. If the lookup
+ * returns a valid ID, we insert that ID into the hash table. Otherwise, we skip adding an
+ * entry into the hash table for the inner column. We can also skip adding any entries
+ * that are outside the range of the outer column.
+ *
+ * Consider a join of the form SELECT x, n FROM (SELECT x, COUNT(*) n FROM t1 GROUP BY x
+ * HAVING n > 10), t2 WHERE t1.x = t2.x; Let the result of the subquery be t1_s.
+ * Due to the HAVING clause, the range of all IDs in t1_s must be less than or equal to
+ * the range of all IDs in t1. Suppose we have an element a in t2.x that is also in
+ * t1_s.x. Then the ID of a must be within the range of t1_s. Therefore it is safe to
+ * ignore any element ID that is not in the dictionary corresponding to t1_s.x or is
+ * outside the range of column t1_s.
+ */
+
+inline int64_t translate_str_id_to_outer_dict(const int64_t elem,
+                                              const int64_t min_elem,
+                                              const int64_t max_elem,
+                                              const void* sd_inner_proxy,
+                                              const void* sd_outer_proxy) {
+  CHECK(sd_outer_proxy);
+  const auto sd_inner_dict_proxy =
+      static_cast<const StringDictionaryProxy*>(sd_inner_proxy);
+  const auto sd_outer_dict_proxy =
+      static_cast<const StringDictionaryProxy*>(sd_outer_proxy);
```

Comment:
`StringDictionaryProxy` class is not declared in this header.

Reply:
done